File: test-zig/module.zig
//! Test guest module for the Zig SDK

const std = @import("std");
const mdb = @import("mdb");

var txn: mdb.Txn = .{ .id = 0 };
var cur: mdb.Cursor = .{ .id = 0 };
var dbi: mdb.DBI = 0;
var k: [16]u8 = undefined;
var v: [16]u8 = undefined;
var statbuf: [48]u8 = undefined;

fn fatal(err: anyerror) noreturn {
    @panic(@errorName(err));
}

export fn memcheck() u32 {
    // No runtime heap allocation in this module; the SDK uses static buffers.
    return 0;
}

export fn begin() void {
    txn = mdb.begin(0) catch |err| fatal(err);
}

export fn beginread() void {
    txn = mdb.begin(mdb.readonly) catch |err| fatal(err);
}

export fn db() void {
    dbi = txn.openDBI("test", mdb.create) catch |err| fatal(err);
}

export fn dbstat() u64 {
    const s = txn.stat(dbi) catch |err| fatal(err);
    s.toBytes(&statbuf);
    return (@as(u64, @intCast(@intFromPtr(&statbuf[0]))) << 32) + statbuf.len;
}

export fn dbdrop() void {
    txn.drop(dbi) catch |err| fatal(err);
}

export fn set() void {
    txn.put(dbi, "a", "1", 0) catch |err| fatal(err);
}

export fn get() void {
    const val = txn.get(dbi, "a") catch |err| fatal(err);
    if (val.len > 1 or val[0] != '1') @panic("wrong value");
}

export fn getmissing() void {
    if (txn.get(dbi, "ddd")) |_| {
        @panic("error not returned");
    } else |err| {
        if (!mdb.isNotFound(err)) fatal(err);
    }
}

export fn del() void {
    txn.del(dbi, "a", "") catch |err| fatal(err);
}

export fn commit() void {
    txn.commit() catch |err| fatal(err);
}

export fn set2() void {
    txn.put(dbi, "b", "2", 0) catch |err| fatal(err);
}

export fn get2() void {
    const val = txn.get(dbi, "b") catch |err| fatal(err);
    if (!std.mem.eql(u8, val, "2")) @panic("wrong value");
}

fn updateFn(t: mdb.Txn) anyerror!void {
    t.put(dbi, "b", "22", 0) catch {};
}

export fn update() void {
    mdb.update(updateFn) catch |err| fatal(err);
}

fn updatefailFn(t: mdb.Txn) anyerror!void {
    t.put(dbi, "b", "222", 0) catch {};
    return error.ICantBelieveYouveDoneThis;
}

export fn updatefail() void {
    if (mdb.update(updatefailFn)) |_| {
        @panic("Error missing");
    } else |_| {}
}

fn viewFn(t: mdb.Txn) anyerror!void {
    const val = try t.get(dbi, "b");
    if (!std.mem.eql(u8, val, "22")) return error.WrongValue;
}

export fn view() void {
    mdb.view(viewFn) catch |err| fatal(err);
}

fn clearFn(t: mdb.Txn) anyerror!void {
    dbi = t.openDBI("test", mdb.create) catch |err| fatal(err);
    return t.drop(dbi);
}

export fn clear() void {
    mdb.update(clearFn) catch |err| fatal(err);
}

fn subPut(t: mdb.Txn) anyerror!void {
    return t.put(dbi, "sub", "txn", 0);
}

fn subFn(t: mdb.Txn) anyerror!void {
    return t.sub(subPut);
}

export fn sub() void {
    mdb.update(subFn) catch |err| fatal(err);
}

fn subPutFail(t: mdb.Txn) anyerror!void {
    t.put(dbi, "sub", "txn", 0) catch {};
    return error.ICantBelieveYouveDoneThis;
}

fn subabortFn(t: mdb.Txn) anyerror!void {
    t.sub(subPutFail) catch {};
}

export fn subabort() void {
    mdb.update(subabortFn) catch |err| fatal(err);
}

fn subDel(t: mdb.Txn) anyerror!void {
    return t.del(dbi, "sub", "");
}

fn subdelFn(t: mdb.Txn) anyerror!void {
    return t.sub(subDel);
}

export fn subdel() void {
    mdb.update(subdelFn) catch |err| fatal(err);
}

export fn stress(limit: u32) void {
    txn = mdb.begin(0) catch |err| fatal(err);
    dbi = txn.openDBI("test", mdb.create) catch |err| fatal(err);
    const n: u64 = limit;
    var i: u64 = 0;
    while (i < n) : (i += 1) {
        std.mem.writeInt(u64, k[0..8], i + 1_000_000_000_000_000, .little);
        std.mem.writeInt(u64, v[0..8], n - i + 1_000_000_000_000_000, .little);
        txn.put(dbi, k[0..8], v[0..8], 0) catch |err| fatal(err);
    }
    txn.commit() catch |err| fatal(err);
}

export fn abort() void {
    txn.abort();
}

export fn cursoropen() void {
    dbi = txn.openDBI("test", mdb.create) catch |err| fatal(err);
    cur = txn.openCursor(dbi) catch |err| fatal(err);
}

export fn cursorfirst() void {
    const e = cur.get("", "", mdb.op_first) catch |err| fatal(err);
    if (!std.mem.eql(u8, e.key, "b")) @panic("wrong key");
    if (!std.mem.eql(u8, e.val, "22")) @panic("wrong value");
}

export fn cursorput() void {
    cur.put("c", "3", 0) catch |err| fatal(err);
}

export fn cursorcurrent() void {
    const e = cur.get("", "", mdb.op_get_current) catch |err| fatal(err);
    if (!std.mem.eql(u8, e.key, "c")) @panic("wrong key");
    if (!std.mem.eql(u8, e.val, "3")) @panic("wrong value");
}

export fn cursornext() void {
    const e = cur.get("", "", mdb.op_next) catch |err| fatal(err);
    if (!std.mem.eql(u8, e.key, "c")) @panic("wrong key");
    if (!std.mem.eql(u8, e.val, "3")) @panic("wrong value");
}

export fn cursordel() void {
    cur.del(mdb.current) catch |err| fatal(err);
}

export fn cursorclose() void {
    cur.close();
}

export fn valptrs2() u64 {
    return (@as(u64, @intCast(@intFromPtr(&k[0]))) << 32) + @as(u64, @intCast(@intFromPtr(&v[0])));
}

export fn setval() void {
    txn.put(dbi, k[0..16], v[0..16], 0) catch |err| fatal(err);
}

File: test-go/module.go
package main

import (
	"encoding/binary"
	"errors"
	"runtime"
	"unsafe"

	"github.com/pantopic/ext-mdb/sdk-go"
)

func main() {}

var txn mdb.Txn
var cur mdb.Cursor
var dbi mdb.DBI
var err error
var k = make([]byte, 16)
var v = make([]byte, 16)
var statbuf = make([]byte, 48)

var m = &runtime.MemStats{}

//export memcheck
func memcheck() (res uint32) {
	runtime.ReadMemStats(m)
	return uint32(m.HeapAlloc)
}

//export begin
func begin() {
	txn, err = mdb.Begin(0)
	if err != nil {
		panic(err)
	}
}

//export beginread
func beginread() {
	txn, err = mdb.Begin(mdb.Readonly)
	if err != nil {
		panic(err)
	}
}

//export db
func db() {
	dbi, err = txn.OpenDBI("test", mdb.Create)
	if err != nil {
		panic(err)
	}
}

//export dbstat
func dbstat() uint64 {
	s, err := txn.Stat(dbi)
	if err != nil {
		panic(err)
	}
	return sliceToPtr(s.ToBytes(statbuf))
}

//export dbdrop
func dbdrop() {
	err := txn.Drop(dbi)
	if err != nil {
		panic(err)
	}
}

//export set
func set() {
	err := txn.Put(dbi, []byte(`a`), []byte(`1`), 0)
	if err != nil {
		panic(err)
	}
}

//export get
func get() {
	v, err := txn.Get(dbi, []byte(`a`), v)
	if err != nil {
		panic(err)
	}
	if len(v) > 1 || v[0] != '1' {
		panic(`wrong value`)
	}
}

//export getmissing
func getmissing() {
	v, err = txn.Get(dbi, []byte(`ddd`), v)
	if err == nil {
		panic(`error not returned`)
	}
	if !mdb.IsNotFound(err) {
		panic(err)
	}
}

//export del
func del() {
	err := txn.Del(dbi, []byte(`a`), nil)
	if err != nil {
		panic(err)
	}
}

//export commit
func commit() {
	err := txn.Commit()
	if err != nil {
		panic(err)
	}
}

//export set2
func set2() {
	err := txn.Put(dbi, []byte(`b`), []byte(`2`), 0)
	if err != nil {
		panic(err)
	}
}

//export get2
func get2() {
	v, err := txn.Get(dbi, []byte(`b`), v)
	if err != nil {
		panic(err)
	}
	if unsafe.String(&v[0], len(v)) != `2` {
		panic(`wrong value`)
	}
}

//export update
func update() {
	if err := mdb.Update(func(txn mdb.Txn) error {
		txn.Put(dbi, []byte(`b`), []byte(`22`), 0)
		return nil
	}); err != nil {
		panic(err)
	}
}

var icbydt = errors.New(`I can't believe you've done this.`)

//export updatefail
func updatefail() {
	if err := mdb.Update(func(txn mdb.Txn) error {
		txn.Put(dbi, []byte(`b`), []byte(`222`), 0)
		return icbydt
	}); err == nil {
		panic(`Error missing`)
	}
}

//export view
func view() {
	err := mdb.View(func(txn mdb.Txn) (err error) {
		v, err = txn.Get(dbi, []byte(`b`), v)
		if err != nil {
			return
		}
		s := unsafe.String(&v[0], len(v))
		if s != `22` {
			err = errors.New("Wrong value: " + string(v) + " != 22")
		}
		return
	})
	if err != nil {
		panic(err)
	}
}

//export clear
func clear() {
	if err := mdb.Update(func(txn mdb.Txn) (err error) {
		dbi, err = txn.OpenDBI("test", mdb.Create)
		if err != nil {
			panic(err)
		}
		return txn.Drop(dbi)
	}); err != nil {
		panic(err)
	}
}

//export sub
func sub() {
	if err := mdb.Update(func(txn mdb.Txn) (err error) {
		return txn.Sub(func(txn mdb.Txn) error {
			return txn.Put(dbi, []byte(`sub`), []byte(`txn`), 0)
		})
	}); err != nil {
		panic(err)
	}
}

//export subabort
func subabort() {
	if err := mdb.Update(func(txn mdb.Txn) (err error) {
		txn.Sub(func(txn mdb.Txn) error {
			txn.Put(dbi, []byte(`sub`), []byte(`txn`), 0)
			return icbydt
		})
		return nil
	}); err != nil {
		panic(err)
	}
}

//export subdel
func subdel() {
	if err := mdb.Update(func(txn mdb.Txn) (err error) {
		return txn.Sub(func(txn mdb.Txn) error {
			return txn.Del(dbi, []byte(`sub`), nil)
		})
	}); err != nil {
		panic(err)
	}
}

//export stress
func stress(limit uint32) {
	txn, err = mdb.Begin(0)
	if err != nil {
		panic(err)
	}
	if dbi, err = txn.OpenDBI("test", mdb.Create); err != nil {
		panic(err)
	}
	n := uint64(limit)
	for i := range n {
		k = binary.LittleEndian.AppendUint64(k[:0], i+1e15)
		v = binary.LittleEndian.AppendUint64(v[:0], n-i+1e15)
		if err := txn.Put(dbi, k, v, 0); err != nil {
			panic(err)
		}
	}
	err := txn.Commit()
	if err != nil {
		panic(err)
	}
}

//export abort
func abort() {
	txn.Abort()
}

//export cursoropen
func cursoropen() {
	if dbi, err = txn.OpenDBI("test", mdb.Create); err != nil {
		panic(err)
	}
	if cur, err = txn.OpenCursor(dbi); err != nil {
		panic(err)
	}
}

//export cursorfirst
func cursorfirst() {
	k, v, err := cur.Get(k, v, mdb.First)
	if err != nil {
		panic(err)
	}
	if unsafe.String(&k[0], len(k)) != `b` {
		panic(`wrong key: ` + string(k))
	}
	if unsafe.String(&v[0], len(v)) != `22` {
		panic(`wrong value: ` + string(v))
	}
}

//export cursorput
func cursorput() {
	err := cur.Put([]byte(`c`), []byte(`3`), 0)
	if err != nil {
		panic(err)
	}
}

//export cursorcurrent
func cursorcurrent() {
	k, v, err := cur.Get(k, v, mdb.GetCurrent)
	if err != nil {
		panic(err)
	}
	if unsafe.String(&k[0], len(k)) != `c` {
		panic(`wrong key: ` + string(k))
	}
	if unsafe.String(&v[0], len(v)) != `3` {
		panic(`wrong value: ` + string(v))
	}
}

//export cursornext
func cursornext() {
	k, v, err := cur.Get(k, v, mdb.Next)
	if err != nil {
		panic(err)
	}
	if unsafe.String(&k[0], len(k)) != `c` {
		panic(`wrong key: ` + string(k))
	}
	if unsafe.String(&v[0], len(v)) != `3` {
		panic(`wrong value: ` + string(v))
	}
}

//export cursordel
func cursordel() {
	err := cur.Del(mdb.Current)
	if err != nil {
		panic(err)
	}
}

//export cursorclose
func cursorclose() {
	cur.Close()
}

//export valptrs2
func valptrs2() uint64 {
	return uint64(uintptr(unsafe.Pointer(&k[0])))<<32 + uint64(uintptr(unsafe.Pointer(&v[0])))
}

//export setval
func setval() {
	if err := txn.Put(dbi, k[:16], v[:16], 0); err != nil {
		panic(err)
	}
}

func sliceToPtr(b []byte) uint64 {
	return uint64(uintptr(unsafe.Pointer(&b[0])))<<32 + uint64(len(b))
}

// Fix for lint rule `unusedfunc`
var _ = memcheck
var _ = begin
var _ = db
var _ = dbstat
var _ = set
var _ = set2
var _ = commit
var _ = abort
var _ = get
var _ = get2
var _ = beginread
var _ = stress
var _ = dbdrop
var _ = del
var _ = cursoropen
var _ = cursorfirst
var _ = cursorput
var _ = cursornext
var _ = cursorcurrent
var _ = cursorclose
var _ = cursordel
var _ = view
var _ = update
var _ = updatefail
var _ = getmissing
var _ = sub
var _ = subabort
var _ = subdel
var _ = clear
var _ = setval
var _ = valptrs2
